[FLINK-20118][file connector] Extend FileSourceTextLinesITCase with JM/TM failover tests

azagrebin opened the pull request:
This PR switches `FileSourceTextLinesITCase` from mini cluster test resource to `TestingMiniCluster` to get manual control over TM start/stop and introduces tests for JM/TM failover in case of streaming continuous file source input.

azagrebin commented:
Note that `FileSourceTextLinesITCase::testContinuousTextFileSourceWithTaskManagerFailover` is currently unstable due to [FLINK-20290](https://issues.apache.org/jira/browse/FLINK-20290).

azagrebin commented:
The last two pushed commits address @StephanEwen's review comments from [WIP branch](https://github.com/azagrebin/flink/commit/e759afef7c321ceda884aab2b8158770a992cd2f#commitcomment-44488706).

XComp reviewed (approved):
LGTM. 👍

StephanEwen commented:
Thanks, @azagrebin this looks good to me.

The setup logic does look a bit complex - I am wondering if it makes sense to have more of the setup logic in the MiniCluster itself? To avoid that future changes to the HAServices need to touch this test class.

What do you think about adding also a failover case for the batch execution? Then we would have everything covered.
We could then rename this test from `FileTextLinesITCase` to a `FileSourceFailoverITCase`.

azagrebin commented:
@XComp @StephanEwen Thanks for the reviews?

> The setup logic does look a bit complex - I am wondering if it makes sense to have more of the setup logic in the
> MiniCluster itself? To avoid that future changes to the HAServices need to touch this test class. 
> (The philosophy being that this class tests DataStream and File Source, so changes not relating to those should ideally not require to touch that class).

We could do this.

Alternatively, `MiniCluster/MiniClusterResource` can be configured by setting a custom `HighAvailabilityServicesFactory` to `HighAvailabilityOptions.HA_MODE` in configuration. The factory can be a separate `TestHAFactory` (we already have some). I would prefer some flexible builder for it because in this test we need something like `TestingEmbeddedHaServices` + `CompletedCheckpointStore`, the former has direct control on leadership and the latter needs to be reset after each test. 

One thing here is that the `TestHAFactory` has to be static to use the config option, hence, the reset of `CompletedCheckpointStore`. So if we could set the factory instance directly to `MiniCluster/MiniClusterConfiguration` that would be more flexible. We could also have a separate issue for this before and after merging this PR (or at least commit).

The reason, why I do not use `MiniClusterResource` in these tests now, is that `MiniCluster` does not expose `terminateTaskManager`. Whether to expose it again may be a separate issue or at least commit. cc @tillrohrmann 

> What do you think about adding also a failover case also for the batch execution? Then we would have everything covered.

I can do this. My thinking was there is less control over concurrency between job run/completion and the failure in batch. This might not trigger the actual test all the time. Alternatively, we could introduce some managed blocking operation in the middle of reading, e.g. with an extra map in the pipeline to sync the execution with the failure. It is more complicated but not random.

> We could then rename this test from `FileTextLinesITCase` to a `FileSourceFailoverITCase`.

I would still keep the same tests w/o failure to keep the coverage, hence, not sure about adding `Failover`. On the other hand, it will be mostly about failover.

StephanEwen commented:
For the batch execution: Not sure over-designing the test is the right way to go here. This stuff tends to be borderline "implementation testing" not "contract testing" and gets expensive in maintenance.
True, even if all slots are used, we would not necessarily get a task failure every time. But it should be enough to cause an unstable test if the batch recovery is broken. 

The main components of Batch recovery, also for the sources, are tested individually in a reliable way. The test here would be for the full scenario, close to the user's perspective.

azagrebin commented:
I added TM failure for batch case.

The JM failure is tricky w/o synchronization with the job execution. The job gets done before JM failure.
We could make `EmbeddedHaServices::getJobManagerLeaderService` public to see whether the JM is still available to restart but this is fragile to concurrency. Hence, I am not sure about the value of this test.

I added JM failure test for batch with the job execution synchronization, abstracting it away, in the last commit. I can remove the last commit if it looks too complicated and there is no better idea how to do the JM failure test for batch in a simpler way.

StephanEwen commented on flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/src/FileSourceTextLinesITCase.java:
```diff
@@ -61,20 +82,83 @@
+	private static CompletedCheckpointStore checkpointStore;
+
+	@BeforeClass
+	public static void setupClass() throws Exception  {
```
Nit: It help reading if the method name is descriptive about what the method does, like `setupMiniCluster()`. The fact that it runs at the beginning of the class is expressed in the annotation already.

StephanEwen commented on flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/src/FileSourceTextLinesITCase.java:
```diff
@@ -61,20 +82,83 @@
+
+	private static Configuration createConfiguration() throws IOException {
+		final Configuration configuration = new Configuration();
+		configuration.set(CheckpointingOptions.STATE_BACKEND, "filesystem");
```
Does it make a difference here which state backend we use, "filesystem" versus the default memory state backend?
Otherwise I would leave this out of here, following the principle that a test should not deal with aspects unrelated to what it actually tests.

azagrebin replied:
I think, we can remove `filesystem`. I think it is just a leftover from my experiments that the checkpoint data stays available between fail-overs.

StephanEwen commented on flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/src/FileSourceTextLinesITCase.java:
```diff
@@ -61,20 +82,83 @@
+	}
+
+	@AfterClass
+	public static void teardownClass() throws Exception {
```
Similar as above, a more descriptive method name would be great.

StephanEwen commented on flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/src/FileSourceTextLinesITCase.java:
```diff
@@ -84,18 +168,33 @@ public void testBoundedTextFileSource() throws Exception {
-		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+		final StreamExecutionEnvironment env = new TestStreamEnvironment(miniCluster, PARALLELISM);
 		env.setParallelism(PARALLELISM);
+		env.enableCheckpointing(10L);
```
I think the bounded case should have explicitly no checkpointing enabled, so that we have recovery that re-executes the fully (goes back to the beginning).

azagrebin replied:
makes sense

StephanEwen commented on flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/src/FileSourceTextLinesITCase.java:
```diff
@@ -308,4 +470,88 @@ private static void writeFileAtomically(
+		}
+	}
+
+	private enum RecordCounterToFail {
```
Can we change this to a regular static class? I though the consensus was to not use the enum pattern any more, because there is not really a problem that it fixes and it causes confusion for many developers.

StephanEwen reviewed (approved):
PR and testing approach generally look good.

There are some minor comments that I can address directly during the merge, if you are okay with all comments.

In the long run, would be nice to have support for these failover scenarios in the MiniCluster directly, so that we don't need to put such complex logic in a test case. Currently this test is prone to cause maintenance issues when components like HAServices or so are evolved.

StephanEwen commented:
Do you want to preserve individual commits, or should (can) I squash them during the merge?

azagrebin commented:
Thanks for the review @StephanEwen 
I addressed comments, the commits can be also just merged into one before merge.

StephanEwen commented:
Thanks, @azagrebin. The test is really good, it helped uncover yet another issue during JM failover (race condition in the coordinator restore).
